planetscale-serverless: allow Client as a valid type

mattrobenolt opened the pull request:
I have no actually knowledge of TypeScript, but this should mitigate the issues in https://github.com/drizzle-team/drizzle-orm/issues/1743

Passing a `Client` instance to the drizzle constructor properly handles connection pooling and safely creating new connections when needed.

`Client` itself supports `execute()` and `transaction()` just like an individual `Connection`, just done in a more correct way.

Related: https://github.com/drizzle-team/drizzle-orm-docs/pull/210

As for documentation, currently for `mysql2`, it's documented both how to use a single connection and a connection pool, so this change is our version of a connection pool. So it might make sense to document both ways? But really, I'd personally believe should always use a `Client` rather than an explicit single `Connection` in all cases. Unlike a traditional MySQL connection, our `Connection` is essentially free and doesn't translate to a new TCP connection. It's just an object to store shared state.

Given my very limited knowledge of TypeScript, I'm not really sure how to verify this change works and passes the type checker, but I read through the code in this driver and it seems like this type union would satisfy all of the uses.

Example script I used to validate behavior:
https://gist.github.com/mattrobenolt/b33e965d285caa6ade5c2484c28d9194

<details>
<summary>Session logs</summary>

```
|INFO| ok method="Execute" query="select 2"              session_id="4Bbn0YabRmQ54TpY4PH2r" client_session=false
|INFO| ok method="Execute" query="select 1"              session_id="LM0chCw2uLqRaVvBH9mjI" client_session=false
|INFO| ok method="Execute" query="BEGIN"                 session_id="XK2sRh5KXx_kMXTMAKoZS" client_session=false
|INFO| ok method="Execute" query="select 1"              session_id="XK2sRh5KXx_kMXTMAKoZS" client_session=true
|INFO| ok method="Execute" query="savepoint sp1"         session_id="XK2sRh5KXx_kMXTMAKoZS" client_session=true
|INFO| ok method="Execute" query="select 2"              session_id="XK2sRh5KXx_kMXTMAKoZS" client_session=true
|INFO| ok method="Execute" query="release savepoint sp1" session_id="XK2sRh5KXx_kMXTMAKoZS" client_session=true
|INFO| ok method="Execute" query="COMMIT"                session_id="XK2sRh5KXx_kMXTMAKoZS" client_session=true
```

From these logs, each `session_id` represents a unique Connection. So we can see that the first two queries each used a new session with no `client_session`, meaning, the driver didn't send along an existing session. It was a new session entirely.

Then the subsequent BEGIN until COMMIT block shares the same `session_id` correctly, and after the first BEGIN, we can see `client_session=true` meaning, we are indeed sending along the existing client session for each query.
</details>

mattrobenolt committed:
planetscale-serverless: allow Client as a valid type

I have no actually knowledge of TypeScript, but this should mitigate the
issues in https://github.com/drizzle-team/drizzle-orm/issues/1743

Passing a `Client` instance to the drizzle constructor properly handles
connection pooling and safely creating new connections when needed.

`Client` itself supports `execute()` and `transaction()` just like an
individual `Connection`, just done in a more correct way.

Related: https://github.com/drizzle-team/drizzle-orm-docs/pull/210

As for documentation, currently for `mysql2`, it's documented both how
to use a single connection and a connection pool, so this change is our
version of a connection pool. So it might make sense to document both
ways? But really, I'd personally believe should always use a `Client`
rather than an explicit single `Connection` in all cases. Unlike a
traditional MySQL connection, our `Connection` is essentially free and
doesn't translate to a new TCP connection. It's just an object to store
shared state.

Given my very limited knowledge of TypeScript, I'm not really sure how
to verify this change works and passes the type checker, but I read
through the code in this driver and it seems like this type union would
satisfy all of the uses.

Example script I used to validate behavior:
https://gist.github.com/mattrobenolt/b33e965d285caa6ade5c2484c28d9194

OmarMcAdam commented:
please see https://github.com/drizzle-team/drizzle-orm/issues/1743#issuecomment-1892860001
you can currently pass in a `Client` but the same issue is still present

mattrobenolt commented:
Can you provide evidence otherwise? I cannot reproduce. If you share code along with ps-http-sim logs (or I can run it myself) I can see what's wrong.

0ximjosh commented:
Confirming this works as expected. For those that want to switch before this gets merged, its just a type fix. In the mean time you can just use a ts-ignore and pass the client object.
```ts
import { env } from "process";
import { Client } from "@planetscale/database";
import { drizzle } from "drizzle-orm/planetscale-serverless";

const client = new Client({
  host: env.DATABASE_HOST,
  username: env.DATABASE_USERNAME,
  password: env.DATABASE_PASSWORD,
});

//@ts-expect-error client should be used here, not connection
export const db = drizzle(client);
```

OmarMcAdam commented:
lemme give this a go
didnt have time to set up full replication for the issue
but I can quickly test this fix

OmarMcAdam commented:
confirmed working 100%
used this repo for testing https://github.com/McPizza0/drizzle-ps-concurrency-test

OmarMcAdam commented:
@AndriiSherman this fixes all the linked issues
I'll open a pr to docs to update the new method

mattrobenolt commented:
@McPizza0 Hi, I appreciate it, but we already have this taken care of.

OmarMcAdam commented:
ah yeah, I see the docs pr now: https://github.com/drizzle-team/drizzle-orm-docs/pull/210
🫡

mattrobenolt committed:
Update tests to use Client rather than a Connection

ayrton commented on drizzle-orm/src/planetscale-serverless/session.ts:
```diff
@@ -16,7 +16,7 @@ import type { RelationalSchemaConfig, TablesRelationalConfig } from '~/relations
 import { type Assume, mapResultRow } from '~/utils.ts';
 
-export type PlanetScaleConnection = Connection;
+export type PlanetScaleConnection = Connection | Client;
```
This feels funny to me I think we might want to inline the unions:

```ts
export type PlanetScaleConnection = Connection

export class PlanetscaleSession<
	TFullSchema extends Record<string, unknown>,
	TSchema extends TablesRelationalConfig,
> extends MySqlSession<PlanetscaleQueryResultHKT, PlanetScalePreparedQueryHKT, TFullSchema, TSchema> {
	private client: Client | Connection | Transaction;

	constructor(
		private baseClient: Client | Connection,
		dialect: MySqlDialect,
		tx: Transaction | undefined,
		private schema: RelationalSchemaConfig<TSchema> | undefined,
		private options: PlanetscaleSessionOptions = {},
	) {
		super(dialect);
		this.client = tx ?? baseClient;
		this.logger = options.logger ?? new NoopLogger();
	}
}
```

It doesn't seem `PlanetScaleConnection` is used internally - is this used by consumers? If not I think we can get rid of it

ayrton replied:
Addressed in 2b1643459d7de9ae13e29b5416d8d5ac3af696cd

AndriiSherman commented:
@mattrobenolt thanks a lot for you PR! We will include those changes in the next release

ayrton committed:
Don't encourage ambigious exported type

ayrton committed:
Enforce the usage of a client instance over a connection

Clients handle connection pooling and safely creates new connections when
needed.

ayrton commented:
Per our discussion on Discord I've updated the PR. We're no longer just updating the types, but are also enforcing correct usage through the `Client`

AndriiSherman committed:
Merge branch 'main' into fix-planetscale-types

ayrton committed:
Fix import

ayrton committed:
Client is not used as just a type anymore

AndriiSherman commented on drizzle-orm/src/planetscale-serverless/driver.ts:
```diff
@@ -22,9 +23,26 @@ export type PlanetScaleDatabase<
+	client: Client,
 	config: DrizzleConfig<TSchema> = {},
 ): PlanetScaleDatabase<TSchema> {
+	if (!(client instanceof Client)) {
```
Please use `is()` from `drizzle-orm` package

https://orm.drizzle.team/docs/goodies#compare-objects-types-instanceof-alternative

```ts
if (!is(client, Client)) {
...
```

ayrton committed:
Codestyle

ayrton committed:
No need to accept connection instances anymore

AndriiSherman committed:
Fix instanceOf issue

AndriiSherman committed:
Upgrade database.js driver for tests

AndriiSherman committed:
Check tests for planetscale

AndriiSherman committed:
Merge branch 'main' into fix-planetscale-types

AndriiSherman committed:
Exclude neon http tests from fork test runner
